### ISSUE TITLE
Paletted Values - Load Color Map and Export Color Map freeze QGIS UI

### ISSUE THREAD

**Issue by ashleysommer** [flathub/org.qgis.qgis#115]
I filed a QGIS issue, but looks like this might be flatpak-related
See corresponding QGIS issue: https://github.com/qgis/QGIS/issues/44630

When I attempt to Load Color Map from file, or Export Color Map to File, the QGIS UI freezes unrecoverable.

#### Repro 1
Add a new Raster layer
Change layer styling to "Paletted/Unique values"
Click the "..." dropdown menu
Choose "Load color map from file"
Choose a valid working TXT color map file, click Open
** QGIS GUI Freezes **

#### Repro 2
Add a new Raster layer
Change layer styling to "Paletted/Unique values"
Click "Classify" to generate unique classifications with random colors
Click the "..." dropdown menu
Choose "Export color map to file"
Enter a file name, click Save.
** QGIS GUI Freezes **

#### Outcome
In both of these cases, the whole QGIS GUI is unresponsive, cannot click any button, cannot close QGIS. CPU usage for the process is reported as 0%, but RAM usage is reported very high (2.8GB used).

### Versions

I've replicated this bug in Flatpack'd QGIS v3.18 and v3.20

**Comment by cpba** [comment]
This is fixed, I'm unable to reproduce it with either method in 3.22/3.16.

It might have been fixed upstream at some point, or maybe when updating some bundled library. The way the gui froze when saving/opening files sounds like an issue with portals.

**Comment by aleixpol** [comment]
That would make sense, closing, please reopen if it's not the case.

**Comment by ashleysommer** [comment]
Hi @cpba and @aleixpol 
I've just updated to Flatpak-QGIS v3.22 (and its dependency flatpack runtimes to latest versions) and I'm still getting this UI Freeze.

I've found it affects more than just the Color Map file picker. That just happens to be where I first noticed it. I actually get the UI freeze on _any_ file open or file save dialog box.

Eg:
1) Layer -> Add Raster Layer -> source (file) -> "..." , to open the file picker.
2) Navigate to any compatible raster file, click open.
3) QGIS UI freezes, file is not opened.

or:
1) Create a new project
2) File -> Save, to open a new save dialog box
3) Enter a file name, hit "Save"
4) QGIS UI freezes, file is not saved.

I've reproduced this on Ubuntu 20.04 (with latest KDE plasma), and KDE Neon distro.

**Comment by cpba** [comment]
I've tried reproducing this with fedora kde on a vm but I can't still get it to happen.

Could you please watch the video I'm attaching and see if I'm doing something wrong?

https://user-images.githubusercontent.com/11727844/151355121-8ed427a8-fc06-4c55-bd0d-db42ab82a5bf.mp4

Maybe ubuntu 20.04 has an old version of the kde portal? I'll try that next

**Comment by cpba** [comment]
@ashleysommer  I'm trying now with kubuntu 20.04, it ships with a two year old flatpak version so I figured that might be the cause.

I've found something that looks like a different bug where the filter is blank and you must click it for it to show any files at all, but I still can't reproduce the UI freeze. Is there some other variable I'm not taking into account?

I'm attaching a video just in case:

https://user-images.githubusercontent.com/11727844/151360746-05c5ca19-3ccb-4849-94b3-c4690dd44677.mp4

**Comment by ashleysommer** [comment]
Thanks for your continued and extended tests to reproduce this issue @cpba 
Your videos posted seem to be doing exactly the right steps, and I can see you are not experiencing the same issue.

With your explanations, I think I worked out what might be my issue.

I am using latest KDE plasma libraries from the KDE Neon repositories (including xdg-desktop-portal-kde v5.23.5) but I'm using the flatpak version that is provided by Ubuntu 20.04 (v1.6.5). So the mismatch between a very new xdg-desktop-portal-kde and a (relatively) old flatpak version might be the cause. I'll see if I can update the flatbak version without breaking Ubuntu, and update this thread with my results.

**Comment by ashleysommer** [comment]
Update. That didn't help. I updated to flatpak v1.12.4 and libflatpak0 v1.12.4, but that didn't help, I still see the same problem.

I will create a fresh install in a VM based on latest KDE Neon, and see if I can replicate the issue inside the VM.

**Comment by ashleysommer** [comment]
I rebooted my Kubuntu system and tried to reproduce the error again, but it worked fine! Just one time. I closed QGIS and re-opened it, (on the cmdline, using `flatpak run org.qgis.qgis`, and the issue was back.

I did notice something. Every time I launch QGIS, I get message that says:
```
Can't get document portal: GDBus.Error:org.freedesktop.DBus.Error.NoReply: Message recipient disconnected from message bus without replying
```
I've tried googling that error, and it seems pretty harmless. I don't have any dbus issues or misconfigurations on my system, so I was ignoring it.

But I noticed, the one time after I rebooted, when the file/open file/save dialogs _did_ work, that dbus message _did not appear_ when I ran QGIS. So it must be somehow related?

**Comment by ashleysommer** [comment]
Ok, I have finally worked it out.

I tried to check the status of `xdg-document-portal` using `systemctl --user status xdg-document-portal` it showed it was not running.

I ran `systemctl --user start xdg-document-portal` but got an error, fuser cannot mount, mount point already exists. So the document portal had died/crashed, and left the mount point in place.

So I ran `umount /run/user/1000/doc` to clear the mount, then ran `systemctl --user start xdg-document-portal` again. This time the document portal service started correctly, and without errors. But I noticed I still got the "Can't get document portal" error when running flatpak QGIS, and the UI freeze issue still occurred.

Then I thought, maybe flatpak needs to start the document portal itself. So I did `systemctl --user stop xdg-document-portal` then `umount /run/user/1000/doc`. Then started flatpak QGIS again. Finally, it worked correctly. I didn't get the document portal error message, and the UI freeze issue is gone.

Note, while running flatpak QGIS (and its working properly) `systemctl --user status xdg-document-portal` shows the service _not_ running, but `/run/user/1000/doc` _is_ mounted. I wonder, can only one subsystem use xdg-document-portal at a time? It flatpak meant to run the document-portal itself? All of these details are way over my head, but I know what to do now when I run into the same issue again in the future.
